File: part2/node/CloudantDialogStore.js
'use strict';

const cloudant = require('cloudant');

/**
 * This class is used to store conversation history in Cloudant.
 */
class CloudantDialogStore {

    /**
     * Creates a new instance of CloudantDialogStore.
     * @param {string} cloudantUrl - The url of the of cloudant instance to connect to
     * @param {string} dbName - The name of the database to use
     */
    constructor(cloudantUrl, dbName) {
        this.cloudant = cloudant({
            url: cloudantUrl,
            plugin:'promises'
        });
        this.dbName = dbName;
        this.db = null;
    }

    /**
     * Creates and initializes the database.
     * @returns {Promise.<null|error>} - Null if fulfilled, or an error if rejected 
     */
    init() {
        console.log('Getting dialog database...');
		return this.cloudant.db.list()
            .then((dbNames) => {
                let exists = false;
                for (let dbName of dbNames) {
                    if (dbName == this.dbName) {
                        exists = true;
                    }
                }
                if (!exists) {
                    console.log(`Creating dialog database ${this.dbName}...`);
                    return this.cloudant.db.create(this.dbName);
                }
                else {
                    return Promise.resolve();
                }
            })
            .then(() => {
                this.db = this.cloudant.db.use(this.dbName);
                return Promise.resolve();
            })
			.then(() => {
                // create the date/userId index
				var index = {
					type: 'json',
					index: {
						fields: ['date','userId']
					}
				};
				return this.db.index(index);
            })
            .then(() => {
                return Promise.resolve();
            });
    }

    /**
     * Adds a new conversation to Cloudant.
     * @param userId - The ID of the user
     * @returns {Promise.<object|error>} - The conversation doc stored in Cloudant if fulfilled, or an error if rejected 
     */
    addConversation(userId) {
        var conversationDoc = {
            userId: userId,
            date: Date.now(),
            dialogs: []
        };
        return this.db.insert(conversationDoc)
            .then((body) => {
                conversationDoc._id = body.id;
                conversationDoc._rev = body.rev;
                return Promise.resolve(conversationDoc);
            });
    }

    /**
     * Adds a new dialog to the conversation.
     * @param conversationId - The ID of the conversation in Cloudant
     * @param dialog - The dialog to add to the conversation
     * @returns {Promise.<object|error>} - The updated conversation doc stored in Cloudant if fulfilled, or an error if rejected 
     */
    addDialog(conversationId, dialog) {
        return this.db.get(conversationId)
            .then((conversationDoc) => {
                conversationDoc.dialogs.push(dialog);
                return this.db.insert(conversationDoc)
                    .then((body) => {
                        conversationDoc._id = body.id;
                        conversationDoc._rev = body.rev;
                        return Promise.resolve(conversationDoc);
                    });
            });
    }
}

module.exports = CloudantDialogStore;

File: part2/python/cloudant_dialog_store.py
import json
import time

from cloudant.query import Query
from cloudant.client import Cloudant


class CloudantDialogStore(object):

    def __init__(self, cloudant_dialogname, cloudant_password, cloudant_url, db_name):
        """
        Creates a new instance of CloudantDialogStore.
        Parameters
        ----------
        cloudant_dialogname - The dialogname for the cloudant instance
        cloudant_password - The password for the cloudant instance
        cloudant_url - The url of the of cloudant instance to connect to
        db_name - The name of the database to use
        """
        if cloudant_url.find('@') > 0:
            prefix = cloudant_url[0:cloudant_url.find('://')+3]
            suffix = cloudant_url[cloudant_url.find('@')+1:]
            cloudant_url = '{}{}'.format(prefix, suffix)
        self.client = Cloudant(cloudant_dialogname, cloudant_password, url=cloudant_url)
        self.db_name = db_name

    def init(self):
        """
        Creates and initializes the database.
        """
        try:
            self.client.connect()
            print('Getting dialog database...')
            if self.db_name not in self.client.all_dbs():
                print('Creating dialog database {}...'.format(self.db_name))
                self.client.create_database(self.db_name)
            else:
                print('Dialog database {} exists.'.format(self.db_name))
        finally:
            self.client.disconnect()

    def add_conversation(self, user_id):
        """
        Adds a new conversation to Cloudant.
        Parameters
        ----------
        user_id - The ID of the user
        """
        try:
            self.client.connect()
            db = self.client[self.db_name]
            conversation_doc = {
                'userId': user_id,
                'date': int(time.time()*1000),
                'dialogs': []
            }
            return db.create_document(conversation_doc)
        finally:
            self.client.disconnect()

    def add_dialog(self, conversation_id, dialog):
        """
        Adds a new dialog to the conversation.
        Parameters
        ----------
        conversation_id - The ID of the conversation in Cloudant
        dialog - The dialog to add to the conversation
        """
        try:
            self.client.connect()
            db = self.client[self.db_name]
            converation_doc = db[conversation_id]
            converation_doc['dialogs'].append(dialog)
            return converation_doc.save()
        finally:
            self.client.disconnect()